sql-plugin/src/main/scala/com/nvidia/spark/rapids/basicPhysicalOperators.scala

```diff
@@ -409,47 +409,53 @@ case class GpuProjectAstExec(
     projectList.collect { case ne: NamedExpression => ne.toAttribute }
   }
 
-  override def internalDoExecuteColumnar() : RDD[ColumnarBatch] = {
+  override def internalDoExecuteColumnar(): RDD[ColumnarBatch] = {
+    child.executeColumnar().mapPartitions(buildRetryableAstIterator)
+  }
+
+  def buildRetryableAstIterator(
+      input: Iterator[ColumnarBatch]): Iterator[ColumnarBatch] with AutoCloseable = {
     val numOutputRows = gpuLongMetric(NUM_OUTPUT_ROWS)
     val numOutputBatches = gpuLongMetric(NUM_OUTPUT_BATCHES)
     val opTime = gpuLongMetric(OP_TIME)
     val boundProjectList = GpuBindReferences.bindGpuReferences(projectList, child.output)
     val outputTypes = output.map(_.dataType).toArray
-    val rdd = child.executeColumnar()
-    rdd.mapPartitions { cbIter =>
-      new Iterator[ColumnarBatch] with AutoCloseable {
-        private[this] var compiledAstExprs =
-          withResource(new NvtxWithMetrics("Compile ASTs", NvtxColor.ORANGE, opTime)) { _ =>
-            boundProjectList.safeMap { expr =>
-              // Use intmax for the left table column count since there's only one input table here.
-              expr.convertToAst(Int.MaxValue).compile()
-            }
+    new Iterator[ColumnarBatch] with AutoCloseable {
+      private[this] var maybeSplittedItr: Iterator[ColumnarBatch] = Iterator.empty
+      private[this] var compiledAstExprs =
+        withResource(new NvtxWithMetrics("Compile ASTs", NvtxColor.ORANGE, opTime)) { _ =>
+          boundProjectList.safeMap { expr =>
+            // Use intmax for the left table column count since there's only one input table here.
+            expr.convertToAst(Int.MaxValue).compile()
           }
+        }
 
-        // Don't install the callback if in a unit test
-        Option(TaskContext.get()).foreach { tc =>
-          onTaskCompletion(tc) {
-            close()
-          }
+      // Don't install the callback if in a unit test
+      Option(TaskContext.get()).foreach { tc =>
+        onTaskCompletion(tc) {
+          close()
         }
+      }
 
-        override def hasNext: Boolean = {
-          if (cbIter.hasNext) {
-            true
-          } else {
-            close()
-            false
-          }
+      override def hasNext: Boolean = maybeSplittedItr.hasNext || {
+        if (input.hasNext) {
+          true
+        } else {
+          close()
+          false
         }
+      }
 
-        override def next(): ColumnarBatch = {
-          withResource(cbIter.next()) { cb =>
-            withResource(new NvtxWithMetrics("Project AST", NvtxColor.CYAN, opTime)) { _ =>
-              numOutputBatches += 1
-              numOutputRows += cb.numRows()
+      override def next(): ColumnarBatch = {
+        if (!maybeSplittedItr.hasNext) {
+          val spillable = SpillableColumnarBatch(
+            input.next(), SpillPriorities.ACTIVE_ON_DECK_PRIORITY)
+          maybeSplittedItr = withRetry(spillable, splitSpillableInHalfByRows) { spillable =>
```

Comment:
nit: Could we add in a comment about how AST does not currently support non-deterministic expressions so we don't have to worry about them.

Reply:
Updated